### PR TITLE
DSG-2802: refactor file upload for reusability

### PR DESCRIPTION
**Pull request by kevincheng0**
# Basics
Place all file upload files into a new document feature library.
Decouple file upload from formly and create a file upload service for reusability

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Styling and layout changes
- [ ] Code style changes (formatting, local variables)
- [x] Refactoring (no functional changes)
- [ ] Build or CI related changes
- [ ] Tests
- [ ] Other... Please describe:

## What changes does this PR introduce?
- Move all document related classes and components into `libs/shared/feature/documents`
- Created new `FileUploadService` to enable reusability for all the file upload and processing code
- Refactor file status display with subjects and observables to reduce reliance on `structureClone`

## Checklist

- [ ] I have updated the documentation accordingly
- [ ] I have confirmed all acceptance criteria defined by the story have been met
- [ ] I have confirmed that Sonar Qube findings for new code are reviewed and fixed, or will be whitelisted

### CODE REVIEW

File: libs/shared/feature/documents/src/lib/components/file-uploader/file-upload-browser.component.ts [hunk 1]
```diff
@@ -0,0 +1,188 @@
+/* eslint-disable @typescript-eslint/member-ordering */
```
Comment (davidyoung77) [inline libs/shared/feature/documents/src/lib/components/file-uploader/file-upload-browser.component.ts:1]: Can we handle the ordering now?

File: libs/shared/feature/documents/src/test-setup.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// @ts-expect-error https://thymikee.github.io/jest-preset-angular/docs/getting-started/test-environment
```
Comment (davidyoung77) [inline libs/shared/feature/documents/src/test-setup.ts:1]: Is this auto generated?

File: libs/shared/feature/documents/src/lib/services/file-upload.service.ts [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+      }),
+      // We handle unsubscribe in 3 ways here because this is a polling observable and we want to complete if this cancels or we navigate away
+      take(1),
+      takeUntil(this._cancelUpload$.pipe(filter(name => name === file.name))),
+      untilDestroyed(this),
+    );
+  }
+
+  /**
+   * TODO: Documentation
```
Comment (davidyoung77) [inline libs/shared/feature/documents/src/lib/services/file-upload.service.ts:206]: Should we take care of this todo now.

File: libs/shared/feature/documents/src/lib/services/file-upload.service.ts
Comment (davidyoung77) [inline libs/shared/feature/documents/src/lib/services/file-upload.service.ts:1]: I'm wondering if it makes sense to combine the file upload service with the document form service, or are they different enough?

**Review by davidyoung77 — COMMENTED** [review]
Looks good, I have some comments

File: libs/shared/feature/documents/src/test-setup.ts (hunk 2 above)
Comment (kevincheng0) [reply libs/shared/feature/documents/src/test-setup.ts:1]: yeah, it was auto generated

File: libs/shared/feature/documents/src/lib/services/file-upload.service.ts
Comment (kevincheng0) [reply libs/shared/feature/documents/src/lib/services/file-upload.service.ts:1]: I believe they are different enough. The file upload service stores and provides state while the document form service doesn't. I think it's better to keep them seperate

**Review by davidyoung77 — APPROVED** [review]
Nice work!

File: libs/shared/feature/documents/src/lib/services/file-upload.service.ts [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+interface ProcessingStatus {
+  isProcessing: boolean;
+}
+
+interface UploadingStatus {
+  isUploading: boolean;
+}
+
+@UntilDestroy()
+@Injectable({ providedIn: 'root' })
```
Comment (davidyoung77) [inline libs/shared/feature/documents/src/lib/services/file-upload.service.ts:22]: This sets the service up as a singleton, you will want to set the service as a provider in the upload component

**Review by davidyoung77 — APPROVED** [review]

**Review by sunnyjiao-nv — APPROVED** [review]

**Pull request merged**